rotkehlchen/chain/ethereum/modules/curve/decoder.py

```diff
@@ -221,6 +229,128 @@ def _decode_curve_deposit_events(
             previous_event = event
         return None, []
 
+    def _decode_curve_trades(
+            self,
+            tx_log: EvmTxReceiptLog,
+            decoded_events: list['EvmEvent'],
+    ) -> tuple[Optional['EvmEvent'], list[ActionItem]]:
+        """Decode curve trades made via single pools or curve swap router
+        First determine:
+        - `spender_address`
+        - `receiver_address`
+        - `sold_token_address`
+        - `bought_token_address`
+        - `raw_sold_amount`
+        - `raw_bought_amount`
+
+        Then create assets if `sold_token_address` and `bought_token_address` were found.
+        Then match and label events.
+        Then reshuffle events to make sure that spend and receive are consecutive.
+        """
+
+        # These are nullable because in case a curve pool is not stored in our cache or if it
+        # is a swap in a metapool (TOKEN_EXCHANGE_UNDERLYING) we will skip token check.
+        sold_token_address: Optional[ChecksumEvmAddress] = None
+        bought_token_address: Optional[ChecksumEvmAddress] = None
+
+        pool_or_router_address: ChecksumEvmAddress
+        if tx_log.topics[0] in (TOKEN_EXCHANGE, TOKEN_EXCHANGE_UNDERLYING):
+            pool_address = tx_log.address
+            pool_or_router_address = pool_address
+            owner = hex_or_bytes_to_address(tx_log.topics[1])
+            spender_address = owner
+            receiver_address = owner
+            sold_token_id = hex_or_bytes_to_int(tx_log.data[:32])
+            raw_sold_amount = hex_or_bytes_to_int(tx_log.data[32:64])
+            bought_token_id = hex_or_bytes_to_int(tx_log.data[64:96])
+            raw_bought_amount = hex_or_bytes_to_int(tx_log.data[96:128])
+            if (
+                tx_log.topics[0] == TOKEN_EXCHANGE and
+                pool_address in self.curve_pools and
+                len(self.curve_pools[pool_address]) > max(sold_token_id, bought_token_id)
```

Comment:
Will add a comment, but this is to make sure that the pool and the tokens of the pool are cached so that we don't get an index error further down the line.